processing/src/main/java/org/apache/druid/query/groupby/DeferExpressionDimensions.java

```diff
@@ -0,0 +1,197 @@
+      if (isInnatelyDeferrable(plan, requiredBindingsList, inspector)) {
+        return false;
+      }
+
+      boolean allNumericInputs = true;
+
+      for (final String requiredBinding : requiredBindingsList) {
+        final ColumnCapabilities capabilities = inspector.getColumnCapabilities(requiredBinding);
+        if (capabilities == null) {
+          return false;
+        }
+
+        allNumericInputs = allNumericInputs && capabilities.isNumeric();
+
+        if (!capabilities.isNumeric() && !capabilities.isDictionaryEncoded().isTrue()) {
+          // Not fixed-width.
+          return false;
+        }
+      }
+
+      return !allNumericInputs || (plan.getOutputType() != null && !plan.getOutputType().isNumeric());
+    }
+  },
+
+  ALWAYS("always") {
+    @Override
+    public boolean useDeferredGroupBySelector(
+        ExpressionPlan plan,
+        List<String> requiredBindingsList,
+        ColumnInspector inspector
+    )
+    {
+      return !isInnatelyDeferrable(plan, requiredBindingsList, inspector);
+    }
+  };
+
+  public static final String JSON_KEY = "deferExpressionDimensions";
+
+  private final String jsonName;
+
+  DeferExpressionDimensions(String jsonName)
+  {
+    this.jsonName = jsonName;
+  }
+
+  @JsonCreator
+  public static DeferExpressionDimensions fromString(final String jsonName)
+  {
+    for (final DeferExpressionDimensions value : values()) {
+      if (value.jsonName.equals(jsonName)) {
+        return value;
+      }
+    }
+
+    throw new IAE("Invalid value[%s] for[%s]", jsonName, JSON_KEY);
+  }
+
+  public abstract boolean useDeferredGroupBySelector(
+      ExpressionPlan plan,
+      List<String> requiredBindingsList,
+      ColumnInspector inspector
+  );
+
+  @Override
+  @JsonValue
+  public String toString()
+  {
+    return jsonName;
+  }
+
+  /**
+   * Whether the given expression can be deferred innately by the selector created by
+   * {@link ExpressionVirtualColumn#makeSingleValueVectorDimensionSelector(DimensionSpec, VectorColumnSelectorFactory)}.
+   *
+   * In this case, all options for this enum return false from
+   * {@link #useDeferredGroupBySelector(ExpressionPlan, List, ColumnInspector)}, because there is no need to defer
+   * redundantly.
+   */
+  private static boolean isInnatelyDeferrable(
+      ExpressionPlan plan,
+      List<String> requiredBindingsList,
+      ColumnInspector inspector
+  )
+  {
+    if (plan.getOutputType() != null
+        && plan.getOutputType().is(ExprType.STRING)
+        && requiredBindingsList.size() <= 1) {
+      for (final String requiredBinding : requiredBindingsList) {
+        final ColumnCapabilities requiredBindingCapabilities = inspector.getColumnCapabilities(requiredBinding);
+
+        if (requiredBindingCapabilities == null
+            || !requiredBindingCapabilities.is(ValueType.STRING)
+            || !requiredBindingCapabilities.isDictionaryEncoded().isTrue()) {
+          return false;
+        }
+      }
+
+      return true;
```

Comment:
I don't think so, because non-unique dictionary (like one that maps both `0` and `1` to the same value `foo`) can still be used for deferred evaluation. The non-uniqueness doesn't hurt, because we will be doing another pass after this that fully groups things.

Here I'm assuming that `isDictionaryEncoded` is not going to be set for string columns that lack any kind of dictionary. Like, I'm assuming that the ones where `DimensionSelector#getValueCardinality` is `-1` would have `isDictionaryEncoded` set to `false` (or at least unknown).

Reply:
> Here I'm assuming that `isDictionaryEncoded` is not going to be set for string columns that lack any kind of dictionary. Like, I'm assuming that the ones where `DimensionSelector#getValueCardinality` is -1 would have `isDictionaryEncoded` set to false (or at least unknown).

This is mainly what i was worried about when i wrote the comment, though looking closer at it i think its probably safe since https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/segment/virtual/ExpressionPlan.java#L279 seems like the only risky place and I believe setting that basically ensures we use the deferred single value selector, so :+1: